man/elpa2_print_kernels.1
.TH "elpa2_print_kernels" 1 "Wed Aug 9 2023" "ELPA" \" -*- nroff -*-
.ad l
.nh
.ss 12 0
.SH NAME
elpa2_print_kernels \- provides information, which ELPA2 kernels are available on this system.

.SH SYNOPSIS
elpa2_print_kernels

.SH Description
Provides information, which ELPA2 kernels are available on this system.
.br
It is possible to configure ELPA2 such, that different compute intensive 'ELPA2 kernels' can be chosen at runtime. The service binary elpa2_print_kernels will query the library and tell whether ELPA2 has been configured in this way, and if this is the case which kernels can be chosen at runtime. It will furthermore detail whether ELPA has been configured with OpenMP support.

.SH Options
none

.SH Author
A. Marek, MPCDF

.SH Reporting bugs
Report bugs to the ELPA mail elpa-library@mpcdf.mpg.de

.SH SEE ALSO
\fBelpa_init\fP(3)\fB elpa_allocate\fP(3)\fB elpa_set\fP(3)\fB elpa_setup\fP(3)\fB elpa_eigenvalues\fP(3)\fB elpa_eigenvectors\fP(3)\fB elpa_cholesky\fP(3)\fB elpa_invert_triangular\fP(3)\fB elpa_solve_tridiagonal\fP(3)\fB elpa_hermitian_multiply\fP(3)\fB elpa_uninit\fP(3)\fB elpa_deallocate\fP(3)



Changelog
Changelog for next release


- not yet decided

Changelog for ELPA 2025.01.001.rc1
- port of tridiagonalization step to GPU: ELPA 1stage/2stage solvers achieve
  speedup of up to 40%
- general multiplication routines for distributed matrices on GPUs

Changelog for ELPA 2024.05.001
- support of ROCM 6.x and preparation for AMD Mi300
- allow internal matrix redistribution if device pointer API is used
- do not try to autotune GPU code paths if no GPUs are available
- implement a patch for a bug in cusolverDnXtrtri_bufferSize for CUDA versions
  < 12.1


Changelog for ELPA 2024.03.001
- fix bug in hermitian_mutliply which had been introduced in 2023.11.001
- PoC RCCL support for AMD GPUs, only for experienced users
- significantly faster cholesky decomposition step
- Automatic setting for cublas caching: with CUDA > 12.x a slow down had been
  observed since cublas assumed problematic caching values
- Autoconf >= 2.71 required for building ELPA

Changelog for ELPA 2023.11.001
- enable gpu-streams per default for NVIDIA and AMD GPUs
- Updated / improved documentation and man pages
- Fixed compilation error on AMD GPUs
- Fixed SVE 256 compute kernels
- Allow (currently in parts of ELPA) to use NVIDIA NCCL for device to device
commpunication
- Speed up of GPU version of hermitian_multiply by up to an factor of 4
- significantly faster full-to-tridiagonal step in ELPA 1stage GPU
- significantly faster ELPA 2stage solver on Intel GPUs
- Consistent enabling/disabling of SKEW_SYMMETRIC in header files
- new setup_gpu API function

Changelog for ELPA 2023.05.001
- added CITATION.cff file
- allow test programs to be run with 1 MPI task
- correct a memory leak in the gpu stream setup
- better handling of GPU BLAS handles
- implement the execution of the AMD HIP code path on NVIDIA GPUs
- implement the execution of the SYCL GPU code path on CPUs (debugging)
- port generalized routines to SYCL GPU
- PoC to use NVIDIA NCCL instead of MPI (not production ready)
- somewhat cleanup of documentation

Changelog for ELPA 2022.11.001
- store GPU setup per ELPA object
- clarify documentation a bit
- add a C++ interface, including an example test program
- fix a few bugs in the C interface for the ELPA solvers
- complete the C API
- make sure that OMP_NUM_THREADS is honoured even if omp_threads is not set
- fix MPI_COMMUNICATORS per ELPA object
- significantly improve the performance of the ELPA band-reduction step of 
  the 2step solver
- fix a few minor bugs in AMD GPU port: is now production ready
- allow to use NVIDIA's CUB implementation; experimental feature
- allow to use AMD's rocsolver library
- implement "HIP to ROCm" layer, in order to be able to run AMD GPU code paths
  on NVIDIA devices
- remove the neccessity to provide the CPP variable

Changelog for ELPA 2022.05.001
- implement OpenMP offloading to GPU for Intel GPU for ELPA 1 and 2 stage (
  except for "step tridi_to_band")
- implement SYCL offloading to Intel GPUs for ELPA 1 and 2 stage
- AMD GPU offload has been tested on Mi200 (also with MPI)
- can use ELPA with one individual "gpu stream" per MPI task (Nvidia and AMD
  only)
- allow steps "cholesky", "invert_trm", and "multiply_ab" to be called
  directly with GPU device pointers
- on error ELPA returns rather than aborting to give controll to calling
  application and to allow for error recovery and/or graceful abort
- allow ELPA to build with OpenMP and GPU
- fix an FPE with the Intel compiler and AVX-512 instructions and optimization
  level > -O2
- better checking of user defined options in configure
 

Changelog for ELPA 2021.11.002
- fix an error when choosing the Nvidia GPU kernel (fallback to CPU might have
  been selected)

Changelog for ELPA 2021.11.001

- support of Nvidia cusolver library to accelerate some routines (needs CUDA >= 11.4)
- experimental Nvidia GPU versions for "elpa_invert_trm" and "elpa_cholesky"
  can be tested by setting elpa_set("gpu_invert_trm",1) and
  elpa_set("gpu_cholesky",1). Is not used otherwise
- BUGFIX: error in resort_ev (also backported to 2021.05.002 and 2020.11.001)
- allow to call ELPA eigenvectors and eigenvalues also with GPU device
  pointers for the input matrix, the vectors of eigenvalues and the output 
  matrix for the eigenvectors
- EXPERIMENTAL feature:g new real GPU kernel for Nvidia A100 (provided by Nvidia): can show a
  performance boost if number of vectors per MPI task is > 20000. Most likely
  most benifit in non-MPI version
- as anounced, droping the legacy interface
- more autotuning features, for example using non blocking MPI collectives
- new version of autotunig avoiding a combinatorial grow of possibilities
  (the old autotune version can be still used if
  elpa%autotune_set_api_version(API_VERSION, error) is set to API_VERSION <
  20211125)

Changelog for ELPA 2021.05.002
- no feature changes
- correct the SO version which was wrong in ELPA 2021.05.001

Changelog for ELPA 2021.05.001

- allow the user to set the mapping of MPI tasks to GPU id per set/get
- experimental feature: port to AMD GPUS, works correctly, performance yet
  unclear; only tested --with-mpi=0
- On request, ELPA can print the pinning of MPI tasks and OpenMP thread
- support for FUGAKU: some minor fix still have to be fixed due to compiler
issues
- BUG FIX: if matrix is already banded, check whether bandwidth >= 2. DO NOT
  ALLOW a bandwidth = 1, since this would imply that the input matrix is
  already diagonal which the ELPA algorithms do not support
- BUG FIX in internal test programs: do not consider a residual of 0.0 to be
  an error
- support for skew-symmetric matrices now enabled by default
- BUG FIX in generalized case: in setups like "mpiexec -np 4 ./validate_real_double_generalized_1stage_random 90 90 45`
- ELPA_SETUPS does now (in case of MPI-runs) check whether the user-provided BLACSGRID is reasonable (i.e. ELPA does 
  _not_ rely anymore that the user does check prior to calling ELPA whether the BLACSGRID is ok) if this check fails 

src/elpa2/elpa2_print_kernels.F90
!    This file is part of ELPA.
!
!    The ELPA library was originally created by the ELPA consortium,
!    consisting of the following organizations:
!
!    - Max Planck Computing and Data Facility (MPCDF), formerly known as
!      Rechenzentrum Garching der Max-Planck-Gesellschaft (RZG),
!    - Bergische Universität Wuppertal, Lehrstuhl für angewandte
!      Informatik,
!    - Technische Universität München, Lehrstuhl für Informatik mit
!      Schwerpunkt Wissenschaftliches Rechnen ,
!    - Fritz-Haber-Institut, Berlin, Abt. Theorie,
!    - Max-Plack-Institut für Mathematik in den Naturwissenschaften,
!      Leipzig, Abt. Komplexe Strukutren in Biologie und Kognition,
!      and
!    - IBM Deutschland GmbH
!
!
!    More information can be found here:
!    http://elpa.mpcdf.mpg.de/
!
!    ELPA is free software: you can redistribute it and/or modify
!    it under the terms of the version 3 of the license of the
!    GNU Lesser General Public License as published by the Free
!    Software Foundation.
!
!    ELPA is distributed in the hope that it will be useful,
!    but WITHOUT ANY WARRANTY; without even the implied warranty of
!    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
!    GNU Lesser General Public License for more details.
!
!    You should have received a copy of the GNU Lesser General Public License
!    along with ELPA.  If not, see <http://www.gnu.org/licenses/>
!
!    ELPA reflects a substantial effort on the part of the original
!    ELPA consortium, and we ask you to respect the spirit of the
!    license that we chose: i.e., please contribute any changes you
!    may have back to the original ELPA library distribution, and keep
!    any derivatives of ELPA under the same license that we chose for
!    the original distribution, the GNU Lesser General Public License.
!
! ELPA2 -- 2-stage solver for ELPA
!
! Copyright of the original code rests with the authors inside the ELPA
! consortium. The copyright of any additional modifications shall rest
! with their original authors, but shall adhere to the licensing terms
! distributed along with the original code in the file "COPYING".
!
! This file was written by A. Marek, MPCDF

#include "config-f90.h"

!> \file print_available_elpa2_kernels.F90
!> \par
!> \brief Provide information which ELPA2 kernels are available on this system
!>
!> \details
!> It is possible to configure ELPA2 such, that different compute intensive
!> "ELPA2 kernels" can be choosen at runtime.
!> The service binary print_available_elpa2_kernels will query the library and tell
!> whether ELPA2 has been configured in this way, and if this is the case which kernels can be
!> choosen at runtime.
!> It will furthermore detail whether ELPA has been configured with OpenMP support
!>
!> Synopsis: print_available_elpa2_kernels
!>
!> \author A. Marek (MPCDF)

program print_available_elpa2_kernels
  use elpa
  use, intrinsic :: iso_c_binding

  implicit none

  integer(kind=c_int) :: i
  class(elpa_t), pointer :: e
  integer :: option, error

  if (elpa_init(CURRENT_API_VERSION) /= ELPA_OK) then
    print *, "Unsupported ELPA API Version"
    stop 1
  endif

  e => elpa_allocate(error)

  print *, "This program will give information on the ELPA2 kernels, "
  print *, "which are available with this library and it will give "
  print *, "information if (and how) the kernels can be choosen at "
  print *, "runtime"
  print *
#ifdef WITH_OPENMP_TRADITIONAL
  print *, " ELPA supports threads: yes"
#else
  print *, " ELPA supports threads: no"
#endif
  print *

  print *, "Information on ELPA2 real case: "
  print *, "=============================== "
#ifdef HAVE_ENVIRONMENT_CHECKING
  print *, " choice via environment variable: yes"
  print *, " environment variable name      : ELPA_DEFAULT_real_kernel"
#else
  print *, " choice via environment variable: no"
#endif
  print *
  print *, " Available real kernels are: "
  print *
  call print_options(e, "real_kernel")
  print *
  print *

  print *, "Information on ELPA2 complex case: "
  print *, "=============================== "
#ifdef HAVE_ENVIRONMENT_CHECKING
  print *, " choice via environment variable: yes"
  print *, " environment variable name      : ELPA_DEFAULT_complex_kernel"
#else
  print *,  " choice via environment variable: no"
#endif
  print *
  print *, " Available complex kernels are: "
  print *
  call print_options(e, "complex_kernel")
  print *
  print *

  call elpa_deallocate(e, error)

  contains

    subroutine print_options(e, KERNEL_KEY)
      class(elpa_t), intent(inout) :: e
      character(len=*), intent(in) :: KERNEL_KEY
      integer                      :: i, kernel,error

      call e%set("solver",ELPA_SOLVER_2STAGE,error)

      do i = 0, elpa_option_cardinality(KERNEL_KEY)
        kernel = elpa_option_enumerate(KERNEL_KEY, i)
        if (elpa_int_value_to_string(KERNEL_KEY, kernel) .eq. "ELPA_2STAGE_COMPLEX_GPU" .or. &
            elpa_int_value_to_string(KERNEL_KEY, kernel) .eq. "ELPA_2STAGE_REAL_GPU") then
          if (e%can_set("gpu",1) == ELPA_OK) then
            call e%set("gpu",1, error)
          endif
        endif 
        if (elpa_int_value_to_string(KERNEL_KEY, kernel) .eq. "ELPA_2STAGE_COMPLEX_NVIDIA_GPU" .or. &
            elpa_int_value_to_string(KERNEL_KEY, kernel) .eq. "ELPA_2STAGE_REAL_NVIDIA_GPU") then
          if (e%can_set("nvidia-gpu",1) == ELPA_OK) then
            call e%set("nvidia-gpu",1, error)
          endif
        endif 
        if (elpa_int_value_to_string(KERNEL_KEY, kernel) .eq. "ELPA_2STAGE_COMPLEX_AMD_GPU" .or. &
            elpa_int_value_to_string(KERNEL_KEY, kernel) .eq. "ELPA_2STAGE_REAL_AMD_GPU") then
          if (e%can_set("amd-gpu",1) == ELPA_OK) then
            call e%set("amd-gpu",1, error)
          endif
        endif 

        if (e%can_set(KERNEL_KEY, kernel) == ELPA_OK) then
          if (elpa_int_value_to_string(KERNEL_KEY, kernel) .eq. "ELPA_2STAGE_COMPLEX_GPU" .or. &
              elpa_int_value_to_string(KERNEL_KEY, kernel) .eq. "ELPA_2STAGE_REAL_GPU" .or. &
              elpa_int_value_to_string(KERNEL_KEY, kernel) .eq. "ELPA_2STAGE_COMPLEX_NVIDIA_GPU" .or. &
              elpa_int_value_to_string(KERNEL_KEY, kernel) .eq. "ELPA_2STAGE_REAL_NVIDIA_GPU" .or. &
              elpa_int_value_to_string(KERNEL_KEY, kernel) .eq. "ELPA_2STAGE_COMPLEX_AMD_GPU" .or. &
              elpa_int_value_to_string(KERNEL_KEY, kernel) .eq. "ELPA_2STAGE_REAL_AMD_GPU" .or. &
              elpa_int_value_to_string(KERNEL_KEY, kernel) .eq. "ELPA_2STAGE_COMPLEX_INTEL_GPU" .or. &
              elpa_int_value_to_string(KERNEL_KEY, kernel) .eq. "ELPA_2STAGE_REAL_INTEL_GPU" ) then
              print *,"  ",elpa_int_value_to_string(KERNEL_KEY, kernel), &
                      "  GPU kernel (might not be usable if no GPUs present on the host)"
          else
            print *, "  ", elpa_int_value_to_string(KERNEL_KEY, kernel)
          endif
        endif
      end do
    end subroutine

end program print_available_elpa2_kernels
